File: python/pondsocket/src/pondsocket/wire.py
from __future__ import annotations

import json
from typing import Any

from pondsocket_common import ValidationError, parse_client_message

from .types import Event


def event_to_wire(event: Event) -> dict[str, Any]:
    return {
        "action": event.action,
        "event": event.event,
        "channelName": event.channel_name,
        "requestId": event.request_id,
        "payload": event.payload if event.payload is not None else {},
    }


def event_to_json(event: Event) -> str:
    return json.dumps(event_to_wire(event), ensure_ascii=False, separators=(",", ":"))


def wire_to_event(data: dict[str, Any]) -> Event:
    msg = parse_client_message(data)
    return Event(
        action=msg.action.value,
        channel_name=msg.channel_name,
        request_id=msg.request_id,
        event=msg.event,
        payload=msg.payload,
    )


def parse_inbound_text(text: str) -> Event:
    try:
        data = json.loads(text)
    except json.JSONDecodeError as e:
        raise ValidationError(f"Invalid JSON: {e.msg}", "wireMessage") from e
    if not isinstance(data, dict):
        raise ValidationError("Expected object", "wireMessage")
    return wire_to_event(data)


def event_to_pubsub_bytes(event: Event) -> bytes:
    data: dict[str, Any] = {
        "action": event.action,
        "event": event.event,
        "channelName": event.channel_name,
        "requestId": event.request_id,
        "payload": event.payload if event.payload is not None else {},
        "nodeId": event.node_id,
    }
    return json.dumps(data, ensure_ascii=False, separators=(",", ":")).encode("utf-8")


def pubsub_bytes_to_event(data: bytes) -> Event | None:
    try:
        text = data.decode("utf-8")
        wire = json.loads(text)
    except (UnicodeDecodeError, json.JSONDecodeError):
        return None
    if not isinstance(wire, dict):
        return None
    action = wire.get("action")
    channel_name = wire.get("channelName")
    request_id = wire.get("requestId")
    event_name = wire.get("event")
    if not isinstance(action, str) or not action:
        return None
    if not isinstance(channel_name, str) or not channel_name:
        return None
    if not isinstance(request_id, str) or not request_id:
        return None
    if not isinstance(event_name, str) or not event_name:
        return None
    payload_raw = wire.get("payload")
    payload: Any = payload_raw if payload_raw is not None else {}
    return Event(
        action=action,
        channel_name=channel_name,
        request_id=request_id,
        event=event_name,
        payload=payload,
        node_id=str(wire.get("nodeId", "")),
    )

File: rust/pondsocket/src/wire.rs
use pondsocket_common::{ValidationError, parse_client_message};
use serde_json::{Map, Value};

use crate::types::Event;

pub fn event_to_json(event: &Event) -> serde_json::Result<String> {
    serde_json::to_string(event)
}

pub fn parse_inbound_text(text: &str) -> Result<Event, ValidationError> {
    let msg = parse_client_message(text)?;
    Ok(Event {
        action: serde_json::to_string(&msg.action)
            .unwrap_or_default()
            .trim_matches('"')
            .to_owned(),
        channel_name: msg.channel_name,
        request_id: msg.request_id,
        event: msg.event,
        payload: Value::Object(msg.payload),
        node_id: String::new(),
        recipients: Vec::new(),
    })
}

pub fn event_to_pubsub_bytes(event: &Event) -> serde_json::Result<Vec<u8>> {
    serde_json::to_vec(event)
}

pub fn pubsub_bytes_to_event(data: &[u8]) -> Option<Event> {
    serde_json::from_slice(data).ok()
}

pub fn object(entries: impl IntoIterator<Item = (impl Into<String>, Value)>) -> Value {
    Value::Object(
        entries
            .into_iter()
            .map(|(k, v)| (k.into(), v))
            .collect::<Map<String, Value>>(),
    )
}
